onnxruntime/core/providers/webgpu/webgpu_context.h

```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  void Flush();
+
+  webgpu::BufferManager& BufferManager() const { return *buffer_mgr_; }
+
+  inline webgpu::ValidationMode ValidationMode() const {
+    return validation_mode_;
+  }
+
+  void StartProfiling();
+  void CollectProfilingData(profiling::Events& events);
+  void EndProfiling(TimePoint, profiling::Events& events, profiling::Events& cached_events);
+
+  Status Run(ComputeContext& context, const ProgramBase& program);
+
+ private:
+  enum class TimestampQueryType {
+    None = 0,
+    InsidePasses,
+    AtPasses
+  };
+
+  WebGpuContext(WGPUInstance instance, WGPUAdapter adapter, WGPUDevice device, webgpu::ValidationMode validation_mode)
+      : query_type_{TimestampQueryType::None}, instance_{instance}, adapter_{adapter}, device_{device}, validation_mode_{validation_mode} {}
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(WebGpuContext);
+
+  std::vector<const char*> GetEnabledAdapterToggles() const;
+  std::vector<const char*> GetEnabledDeviceToggles() const;
+  std::vector<const char*> GetDisabledDeviceToggles() const;
+  std::vector<wgpu::FeatureName> GetAvailableRequiredFeatures(const wgpu::Adapter& adapter) const;
+  wgpu::RequiredLimits GetRequiredLimits(const wgpu::Adapter& adapter) const;
+  void WriteTimestamp(uint32_t query_index);
+
+  TimestampQueryType query_type_;
+  wgpu::QuerySet query_set_;
+  wgpu::Buffer query_resolve_buffer_;
+
+  struct PendingKernelInfo {
+    PendingKernelInfo(std::string_view kernel_name,
+                      std::string_view program_name,
+                      std::string_view cache_key,
+                      const std::vector<ProgramInput>& inputs,
+                      const std::vector<ProgramOutput>& outputs)
+        : name{absl::StrJoin({kernel_name, program_name}, "_")}, cache_key{cache_key}, inputs{inputs}, outputs{outputs} {}
+
+    PendingKernelInfo(PendingKernelInfo&&) = default;
+    PendingKernelInfo& operator=(PendingKernelInfo&&) = default;
+    ORT_DISALLOW_COPY_AND_ASSIGNMENT(PendingKernelInfo);
+
+    std::string name;
+    std::string cache_key;
+    std::vector<ProgramInput> inputs;
+    std::vector<ProgramOutput> outputs;
+  };
+
+  struct PendingQueryInfo {
+    PendingQueryInfo(std::vector<PendingKernelInfo>&& kernels, wgpu::Buffer query_buffer)
+        : kernels{std::move(kernels)}, query_buffer{query_buffer} {}
+
+    PendingQueryInfo(PendingQueryInfo&&) = default;
+    PendingQueryInfo& operator=(PendingQueryInfo&&) = default;
+    ORT_DISALLOW_COPY_AND_ASSIGNMENT(PendingQueryInfo);
+
+    std::vector<PendingKernelInfo> kernels;
+    wgpu::Buffer query_buffer;
+  };
+
+  // info of kernels pending submission for a single batch
+  std::vector<PendingKernelInfo> pending_kernels_;
+  // info of queries pending appending to profiling events
+  std::vector<PendingQueryInfo> pending_queries_;
+
+  uint64_t gpu_timestamp_offset_ = 0;
+  bool is_profiling_ = false;
+
+  std::once_flag init_flag_;
+
+  wgpu::Instance instance_;
```

Comment:
>wgpu::Instance [](http://example.com/codeflow?start=2&length=14)

wgpu::Instance is a ref-counted pointer?

Reply:
Yes it is.